### DISCUSSION TITLE
Testing with Mocha/Chai

### DISCUSSION THREAD

**Discussion by leo-petrucci** [nextauthjs/next-auth#2353] [Help]
### Question 💬

I'm currently trying to test my endpoints with mocha/chai but having some trouble understanding how. I found [this thread with an explanation](https://github.com/nextauthjs/next-auth/issues/30) from some time ago but I haven't been able to get it to work.

As far as I understand it `auth/csrf` returns the cookie I need to authenticate, but I'm still unclear how I can log-in within my code?

The person in the thread seems to be doing:
```
  before(async () => {
    await require('../../../main')
    await User.remove({})
    newUser1 = await (new User(sampleUser1)).save()
    newAdmin = await (new User(sampleAdmin)).save()
  })
```

But doesn't give any explanation on what the code is doing or how to replicate it in other environments.

Is there an official example anywhere on how to get `next-auth` to work with mocha/chai?

### How to reproduce ☕️

My code is pretty much the same as the one in the other thread, but obviously missing the stuff I couldn't understand. Not sure if this helps.

```
import chai, { expect } from "chai";
import chaiHttp from "chai-http";

let csrfToken: string = "";
let agent: ChaiHttp.Agent = {} as ChaiHttp.Agent;

describe("/api/v1.0/users", () => {
  describe("As Logged user", () => {
    before(async () => {
      agent = await chai.request.agent("http://localhost:8000/api");
    });
    it("should get csrfToken", async (done) => {
      await agent
        .get("/auth/csrf")
        .then((res) => {
          expect(res.body).to.be.a("object");
          expect(res.body).to.have.property("csrfToken");
          csrfToken = res.body.csrfToken;
        })
        .catch((err) => {
          throw err;
        });
    });
  });
});
```

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by balazsorban44** [comment]
You shouldn't need to manually test the sign-in flow, that is our job. :)

I would rather try to mock the endpoints.

cc @lluia could we add this to the documentation about testing as it has come up before? 

https://github.com/nextauthjs/next-auth/discussions/2053

**Reply by leo-petrucci** [reply]
> You shouldn't need to manually test the sign-in flow, that is our job. :)

Haha, I know! My main problem is that I need to test permissions on other queries, and for those I need to have the `next-auth` token.

**Reply by balazsorban44** [reply]
you can create an object desired, that useSession would return, containing any permissions.

Check out https://mswjs.io/

or mock the responses from your api as well.

**Reply by leo-petrucci** [reply]
I'm not using `useSession`, at the moment I'm trying to test `getSession` which is protecting my API routes.
I know how to mock hooks in React Testing Library but I have absolutely no idea how to mock a function with Mocha.

I'll update the discussion once I figure it out, thank you!

**Comment by leo-petrucci** [comment]
I figured it out. I suck at non-frontend testing so there's probably better ways.

I'm using Jest (as it's the only test runner I know how to mock external packages with) and chai (as I already have a setup for testing my serverless api).

```js
import chai from "chai";
import chaiHttp from "chai-http";
import { getSession } from "next-auth/client";
// This is the root of my graphql api that uses Apollo Server Micro
import server from "../../pages/api";

let app: any;

chai.use(chaiHttp);

// mocking the getSession function so we can pretend we're logged in
jest.mock("next-auth/client", () => ({
  getSession: () => ({
    user: {
      name: "User",
      email: "test@example.com",
      image: null,
      role: "USER",
    },
    accessToken:
      "xxx",
    expires: "2021-08-11T20:46:37.984Z",
  }),
}));

beforeAll(async () => {
  app = await server;
});

it("gets protected query", async () => {
  const res = await chai
    .request(app)
    .post("/api")
    .send({ query: "{ status { online } }" });
  expect(res.body.status.online).toBe(true);
});
```
